tests/unit/test_charm.py

```diff
@@ -1,326 +1,153 @@
-    #     )
+from pathlib import Path
+from unittest.mock import patch
+
+from scenario import Container, State
+
+from constants import CONTAINER
+
+
+def parse_spark_properties(out: State, tmp_path: Path) -> dict[str, str]:
+
+    spark_properties_path = (
+        out.get_container(CONTAINER)
+        .layers["charm"]
+        .services["history-server"]
+        .environment["SPARK_PROPERTIES_FILE"]
+    )
+
+    file_path = tmp_path / Path(spark_properties_path).relative_to("/opt")
+
+    assert file_path.exists()
+
+    with file_path.open("r") as fid:
+        return dict(
+            row.rsplit("=", maxsplit=1) for line in fid.readlines() if (row := line.strip())
+        )
+
+
+def test_start_history_server(history_server_ctx):
+    state = State(
+        config={},
+        containers=[Container(name=CONTAINER, can_connect=False)],
+    )
+    out = history_server_ctx.run("install", state)
+    assert out.unit_status.name == "maintenance"
+    assert out.unit_status.message == "Waiting for Pebble"
+
+
+def test_pebble_ready(history_server_ctx, history_server_container):
+    state = State(
+        containers=[history_server_container],
+    )
+    out = history_server_ctx.run(history_server_container.pebble_ready_event, state)
+    assert out.unit_status.name == "blocked"
+    assert out.unit_status.message == "Missing S3 relation"
+
+
+@patch("models.S3ConnectionInfo.verify", return_value=True)
+def test_s3_relation_connection_ok(
+    _, tmp_path, history_server_ctx, history_server_container, s3_relation
+):
+    state = State(
+        relations=[s3_relation],
+        containers=[history_server_container],
+    )
+    out = history_server_ctx.run(s3_relation.changed_event, state)
+    assert out.unit_status.name == "active"
+
+    # Check containers modifications
+    assert len(out.get_container(CONTAINER).layers) == 2
+
+    spark_properties = parse_spark_properties(out, tmp_path)
+
+    # Assert one of the keys
+    assert "spark.hadoop.fs.s3a.endpoint" in spark_properties
+    assert (
+        spark_properties["spark.hadoop.fs.s3a.endpoint"] == s3_relation.remote_app_data["endpoint"]
+    )
+
+
+@patch("models.S3ConnectionInfo.verify", return_value=False)
+def test_s3_relation_connection_ko(
+    _, tmp_path, history_server_ctx, history_server_container, s3_relation
+):
+    state = State(
+        relations=[s3_relation],
+        containers=[history_server_container],
+    )
+    out = history_server_ctx.run(s3_relation.changed_event, state)
+    assert out.unit_status.name == "blocked"
+    assert out.unit_status.message == "Invalid S3 credentials"
+
+
+@patch("models.S3ConnectionInfo.verify", return_value=True)
+def test_s3_relation_broken(
+    _, history_server_ctx, history_server_container, s3_relation, tmp_path
+):
+    state = State(
+        relations=[s3_relation],
+        containers=[history_server_container],
+    )
+    out = history_server_ctx.run(
```

Comment:
We also do this sometimes, it's not 'bad practice' but I would make it more explicit and document it better.

```python
state_after_relation_changed = ctx.run(s3.changed_event, initial_state)
state_after_relation_broken = ctx.run(s3.broken_event, state_after_relation_changed)
```

Also maybe you can consider using the `.replace` API to only inject the bits of the state you are interested in, instead of blindly passing a potentially 'too broad' state into the second run call.

For example:
```python
state_after_relation_changed = ctx.run(s3.changed_event, initial_state)
state_after_relation_broken = ctx.run(s3.broken_event, initial_state.replace(containers=state_after_relation_changed.containers)
```

This will ensure that the state is 'cleaner': you only pass through the containers from the relation-changed run, but the rest of the state remains same as the initial one.

Reply:
Uhm, I might prefer option 1. I kinda like that we are achieving a somewhat `black-box` testing, if there is any possibility of doing black box testing at all in charms... :D 

I would also possibly argue that the state is cleaner in this way, since replacing really feels like patching,  possibly creating states you never see in reality